app/scripts/migrations/162.ts

```diff
@@ -0,0 +1,104 @@
+import { hasProperty, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 162;
+
+type TransactionStateEntry = {
+  transactions: unknown[];
+  next: string | null;
+  lastUpdated: number;
+};
+
+type LegacyTransactionsState = {
+  [accountId: string]: TransactionStateEntry;
+};
+
+type NewTransactionsState = {
+  [accountId: string]: {
+    [chainId: string]: TransactionStateEntry;
+  };
+};
+
+/**
+ * This migration transforms the MultichainTransactionsController state structure
+ * to support per-chain transaction storage. It moves transactions from directly
+ * under the account to be nested under the chainId (Solana in this case).
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to disk.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  if (
+    !hasProperty(state, 'MultichainTransactionsController') ||
+    !isObject(state.MultichainTransactionsController)
+  ) {
+    return state;
+  }
+
+  const transactionsController = state.MultichainTransactionsController;
+
+  if (
+    !hasProperty(transactionsController, 'nonEvmTransactions') ||
+    !isObject(transactionsController.nonEvmTransactions)
+  ) {
+    return state;
+  }
+
+  const { nonEvmTransactions } = transactionsController;
+  const newNonEvmTransactions: NewTransactionsState = {};
+
+  // Migrate each account's transactions to the new nested structure
+  for (const [accountId, accountTransactions] of Object.entries(
+    nonEvmTransactions as LegacyTransactionsState,
+  )) {
+    // If the account already has the new structure, meaning the accountTransactions
+    // doesn't have a direct transactions property, instead it has a chainId as a key,
+    // so we can skip it and continue to the next account
+    if (
+      isObject(accountTransactions) &&
+      !Array.isArray(accountTransactions.transactions)
+    ) {
+      newNonEvmTransactions[accountId] =
+        accountTransactions as unknown as NewTransactionsState[string];
+      continue;
+    }
+
+    // Creates the new structure for this account
+    // Since we know the transactions are from Solana, we use the Solana chainId
+    // 'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp' is Solana mainnet (the only supported so far)
+    newNonEvmTransactions[accountId] = {
+      'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp': {
```

Comment:
Nit: I also think having a `const SOLANA_MAINNET = ...` here, would make things clearer (but yes, we have a comment for this actually)